### ISSUE TITLE
Hidden tabs

### ISSUE THREAD

**Issue by jmalenko** [tabkit/tabkit2#71]
I wondered why Session Manager showed that I have 300 tabs, while I saw only around 100 tabs. Session Manager showed tabs (and I ticked them to get them loaded) that were actually not displayed (displayed in tabs) in TabKit.
Recently, I installed the Showcase add-on to see whats going on. Showcase shown all 300 tabs while there were only 100 tabs displayed by TabKit. After clicking (in Showcase) a tab that was displayed in Showcase and that was not displayed in TabKit, TabKit displayed the tab and all also tabs from the same group. The observation is: for a group, all tabs are either hidden or displayed
However, which tabs are displayed after a restart is not deterministic (some groups are hidden, some get hidden after two or three restarts).
I usually have around 30 add-ons. I disabled all of them but TabKit, but the error remained. After also disabling TabKit (and restarting), the hidden tabs remained hidden. As there was no way to check if they are hidden or closed (really removed), I enabled Showcase again. It turned out that they were only hidden. After showing the tabs (in Showcase) and displaying them, they remained visible (for several restarts).
it seems the 200 hidden tabs accumulated in the past 8 months.
I cannot provide steps to reproduce this error. In fact this error may not be related to TabKit. Still, this issue may urge other to check if they observe a similar error.

**Comment by redyoshi49q** [comment]
I experience this particular issue as well.  On session start, there are grouped tabs that are associated with the session but whose group is not represented by any visible tab.  It is possible to use the Omnibar to switch to a tab in such an invisible tab group (if you are specific enough in your entry to have such a tab listed), after which point the whole group becomes visible and remains visible for the remainder of the session.  Tabs do not disappear during a single browser session; they only disappear when the browser restarts after having been closed or having crashed.

I've upgraded TabKit2 to v0.12.1, and while I did previously run the latest Firefox rather than the ESR version, I've since installed the latter and still experience the bug.  Even though I experience this bug regularly in my standard Firefox session, I have yet to successfully reproduce it in a clean Firefox ESR Portable install made for this purpose, which leads me to believe that there's a good chance that either a TabKit2 setting, the presence of another browser extension, or a setting or condtiion outside of TabKit2 is required to cause this bug to occur in TabKit2.

The use of a DOM inspector on the browser's chrome (I use the extensions Element Inspector and Firebug to accomplish this; Firefox's built in DOM inspector doesn't seem to be capable of inspecting the browser itself) reveal that all tabs that aren't visible as per this bug have the hidden="true" property.  This bug seems to exist because all tabs of these hidden groups, including what should be their active tabs, have this property.  Relatedly, an additional way of working around this bug is by changing the DOM data of a tab from each of the tab groups hidden by this bug so that they no longer contain hidden="true"; this causes a tab of each group to appear, which can then be clicked to open that group.



I've noticed that if the first tab of a group is the active tab when the browser crashes or is closed, this bug almost never manifests itself for that group; likewise, if any tab other than the first of a group is the active tab when the browser crashes or is closed, the bug almost always manifests itself.  Even the open group will be affected by this bug if the first tab isn't the active tab of the group (this can result if an ungrouped tab is the active tab of the entire browser), though if the active tab of the entire browser is a grouped tab, that group will always appear when the browser starts regardless of whether the active tab of that group is the first of that group (this is an exception to the first sentence of this paragraph, and the only such exception that I'm aware of for my main Firefox profile in which this bug occurs).

I've also noticed that during the startup of FIrefox, the tabs that become hidden due to this bug seem to disappear seemingly simultaneously with the appearance of the background color for grouped tabs.  A newly started browser seems to spawn the same number of tabs that it had when it closed, each with the label "New Tab".  After that, the tabs acquire labels according to page tiles (while still retaining the correct count of tabs and default background color).  At the moment when the color of the tabs changes to reflect their group affiliation, though, the number of visible tabs (the count of which includes ungrouped tabs, tab groups that are collapsed into a single tab, and grouped tabs belonging to the currently open group) visibly decreases.  Hopefully, this helps you in identifying the segment of code that's relevant to this bug.



...I realize that this is probably an overwhelmingly large info dump, but I hope it's useful nonetheless.

**Comment by PikachuEXE** [comment]
Will try to reproduce this weekend

**Comment by conquerist** [comment]
I don't know if the following is related to this issue:

In my normal browsing, I often close an entire session by checking if I still need a tab, pressing Ctrl-W, and repeating that process until all tabs are gone and Firefox is closed. However, with TabKit, sometimes there will be only one tab remaining, i.e. I expect that closing this remaining tab will close the entire browser. However, after pressing Ctrl-W, instead of the browser closing, a tab from a previous browsing session appears. Pressing Ctrl-W again will then iterate through the entire tab group of the tab from the previous browsing session. After iterating through that tab group, another hidden/forgotten/invisiable tab group may appear, or the browser may close.

I can't remember if I purposefully closed these tab groups in previous sessions, or if they became hidden on a browser restart, as seems to be the case with this issue. Firefox is set to "show my tabs and windows from last time".

I have not been able to replicate this behavior on a clean install though.

**Comment by redyoshi49q** [comment]
I can confirm the behavior that @conquerist has described; while I don't generally close out the tabs of my session individually, I've found in the past that tab groups affected by this bug appear one by one if you try to close out the last visible tab of a window with bugged tab groups.  If you close out all tabs individually, the window will only close once you close out all of the window's tabs, including the ones hidden through the mechanics of this bug.

I'm not entirely sure about my availability yet, but I can try to join the TabKit2 Gitter chat on Saturday and/or Sunday (most likely mid/late evening in CST) if it would help to narrow down this bug.  I won't be available on Friday, though.

**Comment by a-bronx** [comment]
I have the same issue with unloaded tabs remaining hidden after the browser restart. I use TabKit2 and UnloadTab 0.24.7 to unload tabs not visited too long ago.

I've found a workaround for this: if you type a single space in the "Search Tab", all hidden tabs in the current window suddenly reappear, and after you select one of them they remain visible. You should repeat this for every browser window though. And sometimes this trick works only once per session; trying to repeat it may cause more and more unloaded tabs disappearing, so you have to restart the browser again and start again.

Also, if unloaded tabs are grouped and the group is collapsed, everything below the top level will not be saved in bookmarks on a "Bookmark All Tabs" command. Looks like a related issue.

**Comment by PikachuEXE** [comment]
Need someone to reproduce this with Firefox Portable and steps to reproduce it (after I got the Firefox Portable archive)

**Comment by ormembar** [comment]
I have had the same issue. A solution to get the tabs back to visible is using the search tabs.

**Comment by redyoshi49q** [comment]
I had an idea for debugging this recently.  Would it help if I were to:
* make a copy of my Firefox profile (which currently has bugged groups),
* use Firefox's interface to nuke the entirety of the profile's history/cookies/passwords/etc. (for privacy reasons), *except* the extensions list and currently open tabs, and
* send the resulting copy to you somehow?

While I haven't managed to consistently reproduce this bug or make any observations that would help to narrow down the bug's conditions further, this would at least give you an instance of a bugged profile that you could DOM inspect.  I don't know whether this would yield useful information or not, though.

**Comment by PikachuEXE** [comment]
That would be somehow useful (maybe)
Backup your profile first.
You can also try https://addons.mozilla.org/zh-TW/firefox/addon/profileswitcher/, it might have copy function (forgot)

**Comment by redyoshi49q** [comment]
I'm sending you through email a link to the .zip file which contains the profile directory files (scrubbed of history, but not of tabs or extensions, as previously described).  Since Element Inspector is installed, you can DOM edit the browser chrome by right clicking a tab while holding shift.

There are probably a number of bugged groups, but the first such group is the fourth.  The DOM block of its first tab in that group is below:

    <tab class="tabbrowser-tab" crop="end" onerror="this.removeAttribute('image');" fadein="true" context="tabContextMenu" closetabtext="Close Tab" linkedpanel="panel-3-76" label="mff chlorine - Google Search" visibleLabel="mff chlorine - Google Search" hidden="true" minwidth="100" read="true" tabid="1437883331810-77" possibleparent="1437866781284-1" outoforder="true" lastLoadedKey="1437883336505" lastViewedKey="1437883331856" uriKey="google.com.www/https//search?q=mff+chlorine&ie=utf-8&oe=utf-8#q=mff+chlorine&tbs=qdr:m&start=10" groupid=":oG-1437883331810-77" openerGroup=":oG-1437883331810-77" uriGroup=":dG-google:" groupcollapsed="true" image="https://www.google.com/favicon.ico#-moz-resolution=16,16" style="background-image: linear-gradient(rgb(246, 145, 199), rgb(246, 145, 199)) ! important;"></tab>

The group is bugged because every tab of the group has hidden="true" set.

**Comment by PikachuEXE** [comment]
Let me debug it this weekend
I hope you don't have personal stuff in that profile o_0

P.S. Dunno where to archive files like this -3-

**Comment by redyoshi49q** [comment]
I made a second pass at scrubbing the profile; I had apparently forgotten to delete bookmarks or the backups of sessionstore.js (I imagine that Firefox's devs didn't have this particular privacy concern in mind).  If you've already downloaded the file, please redownload it (the link you already have should still work).

I can't think of anything else to scrub; if anything pops out at you, please let me know.

**Comment by PikachuEXE** [comment]
I will let you know if you got any "treasure" left in there XD

**Comment by PikachuEXE** [comment]
Please try this:
https://bintray.com/tabkit/tabkit2/firefox_extension_package/0.12.3.alpha-1/view

The bug was caused by an old workaround for another issue, but I can't remember what that issue was.
Please test it for 1~2 weeks and report any new issue introduced in this version.

**Comment by redyoshi49q** [comment]
Upon installing the extension and restarting, my session went from having only 1 non-bugged group to having *no* bugged groups.  I think you nailed the fix; I'll let you know if I notice any side effects.

(Thank you so, *so* much, by the way.)

**Comment by redyoshi49q** [comment]
After using the alpha build for a while, I think I found the regression caused by removing your workaround.  If a grouped tab is the active tab upon closing the browser, that tab's group is collapsed during the browser's next start (all groups, including the one that should be active, are visible but collapsed).  This behavior wasn't present in the version of the plugin that I was using previously (0.12.1), and this collapsing seems to happen at the same time of the browser's startup that tab groups were disappearing in the prior versions of the plugin.  Note that my standard use of the browser only uses a portion of Tabkit's range of functionality, and there may be other bugs related to functionality I don't use (for example, I use multi row tabs on a horizontal tab bar on top, and therefore can't attest to the behavior of tree style tab groups on a vertical tab bar).

It's a rather minor bug, though, and it's far less breaking than the bug that caused tab groups to disappear.

**Comment by PikachuEXE** [comment]
If it's a trade of having a new minor bug for fixing a major bug. 
I think it's worth it.
But I will try to reproduce that regression in vertical mode to see if it's fixable / serious.
I don't use collapsed groups myself :P

**Comment by PikachuEXE** [comment]
I can't reproduce it
Maybe you are using a different version of Firefox?

## Multi-row

### Before restart
![before_restart_multi_row](https://cloud.githubusercontent.com/assets/1018543/9565599/359dc962-4f0c-11e5-989c-a70e541ed5d5.png)

### After restart
![after_restart_multi_row](https://cloud.githubusercontent.com/assets/1018543/9565600/417f97ce-4f0c-11e5-9cc8-58174c0260ae.png)

## Tree View

### Before restart
![before_restart](https://cloud.githubusercontent.com/assets/1018543/9565602/649fca08-4f0c-11e5-8244-e0faaf651de8.png)

### After restart
![after_restart](https://cloud.githubusercontent.com/assets/1018543/9565603/6e6191c0-4f0c-11e5-81bf-e4931219cbe9.png)

## My options
![grouping_options](https://cloud.githubusercontent.com/assets/1018543/9565605/84c11972-4f0c-11e5-81be-aa97f8a80f34.png)

**Comment by redyoshi49q** [comment]
I did my own testing, using the Firefox profile that I sent you as my starting point.  I did the following things:

* install alpha tabkit on profile
* restart
* select first tab of first group (gmail)
* restart (bug was **not** reproduced)
* select third group
* select second group
* select XKCD tab of second group
* restart (bug **was** reproduced)
* select first group
* select third tab of first group
* restart (bug **was** reproduced)
* select fourth group
* select last tab of fourth group
* select first tab of fourth group
* restart (bug was **not** reproduced)

It appears that if the bug does not occur if the active tab is the first of a tab group (which I imagine is why you didn't reproduce it in your own test).  Also worth noting is that sometimes (including both reproductions above, as they occurred on my end), only some tabs of the group will disappear (for instance, on the first reproduction listed above, the active tab and 3 others in a group of roughly 20 tabs were visible).

My initial description of the bug came from Firefox ESR 38.2.1, and the Firefox Portable that the testing profile was on was ESR version 31.5.0.

...Maybe I should have filed a separate issue for this?

**Comment by redyoshi49q** [comment]
...And just now, I restarted my browser with the first tab of the first group as the active tab; the bug *did* manifest itself, which means that the evidence I have recently presented now has a counterexample to its deduction.  I now have no idea what the nature of the presumed extra criterion to reproduce this bug is.

**Comment by PikachuEXE** [comment]
Maybe changing some preference will make the bug disappear again?

**Comment by redyoshi49q** [comment]
I'm not sure what settings to change.  Most of my settings were the same as yours, and the ones that aren't in particular don't seem remotely related.

In any case, I managed to reproduce the bug on a clean environment (with the alpha TK2, classic theme restorer extension, and nothing else).  Since it's less of a privacy concern than the prior time I shared a profile directory (again, clean Firefox profile), I'll link it here.

https://www.dropbox.com/s/y61mycpx7mw3vnq/profile_2.zip

**Comment by barefootcoder** [comment]
Okay, so just summing up the conversation from #92:
* The fix for this issue (which appeared in 0.12.3) apparently caused the #92 issue.
* Fixing #92, in 0.12.4, now causes this issue (#71) to reappear.
* To be fair, #92 appears to be FF-version-specific: I noted that I had two computers, both running 0.12.3, but only one (running FF 37.0.2) had the #92 bug; the other (running FF 35.0.1) did not.  Thus, if you are running an older version of FF and this issue (#71) really bugs you, you may be perfectly happy with 0.12.3.

I've never purposefully tried to reproduce the hidden-tabs issue, but I can't think of a single time I've restarted FF when it didn't occur (outside of when I had 0.12.3, which of course fixed the problem).  This may very well be because my sessions commonly include several windows, each containing lots of tabs and several tabgroups.  My workaround has always been to enable Panorama with Tab Kit and just click on a tab from each group in there, which makes them all visible again.

I'm being forced to upgrade my older FF (the 35.0.1 version) due to a different plugin problem, so I will lose access to that version, but I will keep my 37.0.2 version where it is.  Please let me know if I can provide any further info or testing.

**Comment by barefootcoder** [comment]
Just upgraded FF to 43.0 on this machine and I _did_ experience this bug on restart.  So if this bug is FF-version-specific, I have yet to find the version that it doesn't happen on.  Just wanted to throw that out there.

**Comment by a-bronx** [comment]
I have two machines with the same version of FF (42.0) and almost identical set of plugins, and #92  have been seen only on one of them. So the #92 looks related with some other factor -- may be with the profile itself, or with stale session data.

**Comment by redyoshi49q** [comment]
...So... I had been meaning to mention in here that I had started to experience bug #71 again, but it seems like you guys have things figured out already.  I'm now running Tab Kit 0.12.4 (I was apparently auto-updated) and Firefox ESR 38.5.2.  I agree with the consensus that #92 is a worse bug than #71, since there are various workarounds for #71 that'll last for the duration of the running session.

I have to say, these bugs are rather strange in the behavior that they exhibit.

**Comment by M3Rocket** [comment]
I believe this issue is fixed as indicated in a related problem in issue https://github.com/tabkit/tabkit2/issues/92

I vote for closing this issue.

**Comment by PikachuEXE** [comment]
OK
Finally fixed (although I don't really know what happened :P )

**Comment by redyoshi49q** [comment]
I can confirm that Tabkit 0.12.9 fixes this issue on Firefox 45.1.0 ESR.  I had bugged tab groups in my session before upgrading both Tabkit and Firefox, and upgrading both to the aforementioned version numbers caused those groups to appear normally.

**Comment by mattehicks** [comment]
I'm having a problem with FF50, which is probably not supported, but nonetheless it works fine- except that when TabsKit is enabled, I cannot access my 'Addons' menu. Nothing happens when I click the icon.  Any clue how to get this back to normal?

**Comment by PikachuEXE** [comment]
@mattehicks I think it might be a new issue which maybe for new version only
And not the same as this issue
Probably they changed behaviour/interface and TabKit throw an error when trying to do something

I have no idea until I start development for ESR
